[libc][math][c23] Add fabsf16 C23 math function

overmighty opened the pull request:
cc @lntue

overmighty committed:
[libc][math][c23] Add fabsf16 C23 math function

overmighty commented on libc/include/llvm-libc-types/float16.h:
```diff
@@ -0,0 +1,38 @@
+#include "src/__support/macros/properties/compiler.h"
+#include "src/__support/macros/properties/cpu_features.h"
+
+#if defined(LIBC_TARGET_ARCH_IS_X86_64) && defined(LIBC_TARGET_CPU_HAS_SSE2)
+#if (defined(LIBC_COMPILER_CLANG_VER) && (LIBC_COMPILER_CLANG_VER >= 1500)) || \
+    (defined(LIBC_COMPILER_GCC_VER) && (LIBC_COMPILER_GCC_VER >= 1201))
+#define LIBC_TYPES_HAS_FLOAT16
+using float16 = _Float16;
+#endif
+#endif
+#if defined(LIBC_TARGET_ARCH_IS_AARCH64)
+#if (defined(LIBC_COMPILER_CLANG_VER) && (LIBC_COMPILER_CLANG_VER >= 900)) ||  \
+    (defined(LIBC_COMPILER_GCC_VER) && (LIBC_COMPILER_GCC_VER >= 1301))
+#define LIBC_TYPES_HAS_FLOAT16
+using float16 = _Float16;
+#endif
+#endif
+#if defined(LIBC_TARGET_ARCH_IS_ANY_RISCV)
+#if (defined(LIBC_COMPILER_CLANG_VER) && (LIBC_COMPILER_CLANG_VER >= 1300)) || \
+    (defined(LIBC_COMPILER_GCC_VER) && (LIBC_COMPILER_GCC_VER >= 1301))
+#define LIBC_TYPES_HAS_FLOAT16
+using float16 = _Float16;
+#endif
+#endif
```
Moved from libc/src/__support/macros/properties/types.h.

We talked about just checking for `__FLT16_MANT_DIG__` but GCC < 13.1 defines it without supporting `_Float16` literals: https://godbolt.org/z/Pn9Ph6WPG.

lntue replied:
GCC 12 does support `_Float16` and it's the only one that does not support `f16` literal.  This can be checked with `defined(__FLT16_IS_IEC_60559__) && (__FLT16_IS_IEC_60559__ == 2)`.

overmighty replied:
GCC >= 12.2 and < 13.1 also defines `__FLT16_MANT_DIG__` without supporting `_Float16` at all when compiling for AArch64. See https://godbolt.org/z/1d9E39fod.

Maybe we should just check for GCC 13.1 directly since `__FLT16_MANT_DIG__` is reliable on Clang:

```cpp
#if defined(__FLT16_MANT_DIG__) && (!defined(LIBC_COMPILER_IS_GCC) || LIBC_COMPILER_GCC_VER >= 1301)
#define LIBC_TYPES_HAS_FLOAT16
using float16 = _Float16;
#endif
```

lntue replied:
SGTM.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Fix usage of internal float16 alias in generated headers.

overmighty commented on libc/spec/stdc.td:
```diff
@@ -378,6 +378,7 @@ def StdC : StandardSpec<"stdc"> {
       [
           NamedType<"float_t">,
           NamedType<"double_t">,
+          Float16Type,
           NamedType<"float128">,
       ],
```
Why weren't the type records from spec.td being reused here? E.g., why did we have `NamedType<"float128">` instead of `Float128Type`?

overmighty reviewed (commented):
Should we `#include "src/__support/macros/properties/types.h"` for `float16`, like we do for `float128`, or should we `#include "llvm-libc-types/float16.h"`?

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Add check mark for fabsf16 in docs.

nickdesaulniers commented:
> Should we `#include "src/__support/macros/properties/types.h"` for `float16`, like we do for `float128`, or should we `#include "llvm-libc-types/float16.h"`?

I think it's nice to be as specific as possible. To me, long include lists that are precise are a feature, not a bug.

nickdesaulniers commented on libc/src/__support/CPP/type_traits/is_floating_point.h:
```diff
@@ -24,7 +24,14 @@ template <typename T> struct is_floating_point {
 
 public:
-#if defined(LIBC_TYPES_HAS_FLOAT128)
+#if defined(LIBC_TYPES_HAS_FLOAT16) && defined(LIBC_TYPES_HAS_FLOAT128)
+  LIBC_INLINE_VAR static constexpr bool value =
+      __is_unqualified_any_of<T, float, double, long double, float16,
+                              float128>();
+#elif defined(LIBC_TYPES_HAS_FLOAT16)
+  LIBC_INLINE_VAR static constexpr bool value =
+      __is_unqualified_any_of<T, float, double, long double, float16>();
+#elif defined(LIBC_TYPES_HAS_FLOAT128)
   LIBC_INLINE_VAR static constexpr bool value =
       __is_unqualified_any_of<T, float, double, long double, float128>();
```
```suggestion
  LIBC_INLINE_VAR static constexpr bool value =
      __is_unqualified_any_of<T, float, double, long double
#ifdef LIBC_TYPES_HAS_FLOAT16
      , float16
#endif
#ifdef LIBC_TYPES_HAS_FLOAT128
      , float128
#endif
  >();
```

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Fix missing include in generated math.h public header.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Replace includes for float16 alias with "include/llvm-libc-types/float16.h".

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Simplify cpp::is_floating_point.

overmighty commented on libc/spec/stdc.td:
```diff
@@ -378,6 +378,7 @@ def StdC : StandardSpec<"stdc"> {
       [
           NamedType<"float_t">,
           NamedType<"double_t">,
+          NamedType<"float16">,
```
We actually use `_Float16` (`Float16Type`) in the public header, but this needs to match the name of the header in llvm-libc-types for libc-hdrgen to generate the #include line.

Or should we split the `LIBC_TYPES_HAS_FLOAT16` macro and the `float16` type alias into 2 separate headers, as we said we'd only use `float16` internally?

lntue replied:
I think that's fine, you can add some comments to explain the discrepancy here.  Maybe we can update this part when the new header gen is implemented? @michaelrj-google

overmighty replied:
If we move the definition of `float16` to libc/src/__support/macros/properties/types.h, then we could just rename the header that defines `LIBC_TYPES_HAS_FLOAT16` to _Float16.h and we wouldn't have this discrepancy anymore.

overmighty replied:
I see, thanks.

overmighty replied:
If we only define `LIBC_TYPES_HAS_FLOAT16` in the public header, does it still make it sense to have it in the llvm-libc-types directory? Each header in that directory defines a type or has a `typedef`.

Also, I just noticed that I copied the C++ `using` statement to the public header which needs to compile as C. We don't seem to have any tests that use our public headers from C, except libc/test/include/stdbit_test.c.

lntue replied:
Yes, we still need the macro definition in the public header, so that it can be used to guard `_Float16` functions in the generated `math.h`.

overmighty replied:
I wasn't suggesting to move it out of the whole libc/include directory for public headers, but the llvm-libc-types subdirectory currently only stores headers that define types or `typedef`s. However, I'll leave it in llvm-libc-types for now as moving it to another subdirectory would require changes to llvm-hdrgen to generate the proper include.

lntue replied:
OTOH, I think we can just move the `LIBC_TYPES_HAS_FLOAT16` macro definition to `llvm-libc-macros/float16-macros.h`, directly include it in `math.h.def` and don't have to add it here.

overmighty replied:
Oops, didn't see your comment.

overmighty replied:
Just noticed I somehow didn't delete llvm-libc-types/float16.h.

overmighty commented on libc/config/linux/api.td:
```diff
@@ -60,7 +60,7 @@ def IntTypesAPI : PublicAPI<"inttypes.h"> {
 
 def MathAPI : PublicAPI<"math.h"> {
-  let Types = ["double_t", "float_t", "float128"];
+  let Types = ["double_t", "float_t", "float16", "float128"];
```
See comment in stdc.td.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Simplify check for _Float16 support.

overmighty commented on libc/test/src/math/smoke/CMakeLists.txt:
```diff
@@ -92,6 +92,19 @@ add_fp_unittest(
+    FAbsTest.h
+  DEPENDS
+    libc.src.math.fabsf16
+    libc.src.__support.FPUtil.fp_bits
```
Copied from the surrounding test target definitions, but I don't see why `fp_bits` is being added as a dependency for `fabs*` tests.

lntue replied:
Feel free to remove the dependency if `FAbsTest.h` does not use it.

lntue commented on libc/include/llvm-libc-types/float16.h:
```diff
@@ -0,0 +1,20 @@
+#if defined(__FLT16_MANT_DIG__) &&                                             \
+    (!defined(LIBC_COMPILER_IS_GCC) || LIBC_COMPILER_GCC_VER >= 1301)
+#define LIBC_TYPES_HAS_FLOAT16
+using float16 = _Float16;
```
We shouldn't export `float16` symbol in the public header here.

lntue commented on libc/src/__support/macros/properties/types.h:
```diff
@@ -39,28 +40,8 @@
-#endif
-#endif
+// LIBC_TYPES_HAS_FLOAT16 and 'float16' type are provided by
+// "include/llvm-libc-types/float16.h"
```
We should define the `float16` alias here instead.

overmighty commented on libc/include/llvm-libc-types/float16.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef LLVM_LIBC_TYPES_FLOAT16_H
+#define LLVM_LIBC_TYPES_FLOAT16_H
+
+#include "src/__support/macros/properties/compiler.h"
```
I just realized, this is an internal header that we can't include in a public one, right? It isn't copied to the build directory.

lntue replied:
yes, you cannot include anything in the `src` folder from here.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Fix llvm-libc-types/float16.h including internal header and exposing internal C++ using float16 stmt.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Remove unused direct dependency from fabsf16_test target.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Move _Float16 detection header to llvm-libc-macros/float16-macros.h.

overmighty commented:
Rebased and fixed the merge conflict.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Enable fabsf16 entrypoint on linux/aarch64.

overmighty committed:
fixup! [libc][math][c23] Add fabsf16 C23 math function

Delete libc/include/llvm-libc-types/float16.h.
